java/ql/src/semmle/code/java/frameworks/guava/IO.qll

```diff
@@ -0,0 +1,66 @@
+/** Definitions of taint steps in the IO packae of the Guava framework */
+
+import java
+private import semmle.code.java.dataflow.ExternalFlow
+
+private class GuavaIoCsv extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;inputspec;outputspec;kind",
+        "com.google.common.io;BaseEncoding;true;decode;(CharSequence);;Argument[0];ReturnValue;taint",
+        "com.google.common.io;BaseEncoding;true;decodingStream;(Reader);;Argument[0];ReturnValue;taint",
+        "com.google.common.io;BaseEncoding;true;decodingSource;(CharSource);;Argument[0];ReturnValue;taint",
+        "com.google.common.io;BaseEncoding;true;encode;(byte[]);;Argument[0];ReturnValue;taint",
+        "com.google.common.io;BaseEncoding;true;encode;(byte[],int,int);;Argument[0];ReturnValue;taint",
+        "com.google.common.io;BaseEncoding;true;withSeparator;(String,int);;Argument[0];ReturnValue;taint",
```

Comment:
Can taint be modelled backwards?
In that case should this also match `encodingSink` and `encodingStream` where the return value taints the argument?

Reply:
There are a few special cases that model backwards taint steps, however there is no general mechanism for it, as it would likely lead to false positives from "time-travel flow", in cases like the following:
```
x = new StringWriter();
y = enc.encodingStream(x)
sink(x.toString());
y.write(taint());
```
If the writes to `y` are propegated backwards to `x`, then we get spurious flow to sink. 
So, the question of whether to include things like `encodingStream` (in general, wrappers around an output stream - of which there are several in the standard library that we don't model for simlar reasons) amounts to whether the additional effort to implement a special-case backwards flow step plus the potential for these kinds of FPs is worth the few additional results it could get; which I don't think it is.